shell/common/shell.cc

```diff
@@ -372,6 +371,71 @@ void Shell::NotifyLowMemoryWarning() const {
   // to purge them.
 }
 
+void Shell::RunEngine(RunConfiguration run_configuration) {
+  RunEngine(std::move(run_configuration), nullptr);
+}
+
+void Shell::RunEngine(RunConfiguration run_configuration,
+                      std::function<void(Engine::RunStatus)> result_callback) {
```

Comment:
I think it would be better to have the result callback be executed on the calling task runner (the platform task runner in this case). Since doing the re-threading at the each possible call-site is tedious, you can wrap the same in  and closure that does the re-threading and then pass that around as normal. See the image decoder for [a use of this pattern](https://github.com/flutter/engine/blob/e77e4f0e051cb53754814ffe33df291a551bdba3/lib/ui/painting/image_decoder.cc#L200).

On a related note, since forgetting to invoke the closure in the error case is possible, maybe we should have an RAII wrapper for the same. [I have a prototype for this in a different patch that is not in yet](https://github.com/flutter/engine/pull/10195/files#diff-f7bab70419be393edf76223077e9b656R32).

That way, you can wrap the callback in a closure that does the re-threading and put it in that RAII wrapper. You could then invoke the same only in the success case and ensure that all erroneous cases invoke the callback and on the right thread.

Something like the following should work for now:

```
auto result = [runner = task_runner_.GetPlatformTaskRunner(), result_callback]
    (Engine::RunStatus status){
  if (!result_callback) {
    return;
  }
  runner->PostTask([result_callback, status](){ result_callback(status); });
};

// Pass result around and invoke it directly (no null check is necessary as it is performed in the wrapper).
// In the future, result can be stuffed inside an AutoFireClosure.
```